publish.js

```diff
@@ -0,0 +1,270 @@
+    }
+};
+
+/**
+ * Tags the specified version with the specified dist-tag
+ * If it is a test run, the tag command will be output to the console.
+ *
+ * @param isTest {Boolean} - indicates if this is a test run or not
+ * @param version {String} - a string idicating which version to tag
+ * @param tag {String} - the dist-tag to apply
+ * @param options {Object} - e.g. {"distTagCmd": "npm dist-tag add infusion@${version} ${tag}"}
+ */
+publish.tag = function (isTest, version, tag, options) {
+    var cmdTemplate = options.distTagCmd || defaults.distTagCmd;
+    var cmdStr = es6Template(cmdTemplate, {
+        version: version,
+        tag: tag
+    });
+    if (isTest) {
+        publish.log("tag command: " + cmdStr);
+    } else {
+        publish.execSync(cmdStr);
+    }
+};
+
+/**
+ * Resets the current workspace.
+ * This will clear out any git tracked changes.
+ *
+ * Used internally to reset version number changes in package.json
+ * @param options {Object} - e.g. {"cleanCmd": "git reset HEAD --hard"}
+ */
+publish.clean = function (options) {
+    var cmdStr = options.cleanCmd || defaults.cleanCmd;
+    publish.execSync(cmdStr);
+};
+
+/**
+ * Publishes a development build.
+ * This creates a release named after the version, but with the build stamp
+ * appended to the end. By default this will create a release with version
+ * X.x.x-prerelease.yyyymmddThhmmssZ.shortHash where x.x.x-prerelease is sourced
+ * from the version number in the package.json file and the build id
+ * (yyyymmddThhmmssZ.shortHash) is generated based on the latest commit.
+ *
+ * @param isTest {Boolean} - indicates if this is a test run
+ * @param options {Object} - see defaultOptions in package.json for possible values
+ */
+publish.dev = function (isTest, options) {
+    var opts = extend(true, {}, defaults, options);
+
+    // Ensure no uncommitted changes
+    publish.checkChanges(opts);
+
+    var devVersion = publish.getDevVersion(opts);
+
+    // set the version number
+    publish.setVersion(devVersion, opts);
+
+    publish.pubImpl(isTest, opts);
+    publish.tag(isTest, devVersion, opts.devTag, opts);
+
+    // cleanup changes
+    publish.clean(opts);
+};
+
+/**
+ * Publishes a release build.
+ * This creates a release named after the version in the package.json file.
+ * It will not increase the version number, this must be done separately.
+ *
+ * @param isTest {Boolean} - indicates if this is a test run
+ * @param options {Object} - see defaultOptions in package.json for possible values
+ */
+publish.release = function (isTest, options) {
+    var opts = extend(true, {}, defaults, options);
+
+    // Ensure no uncommitted changes
+    publish.checkChanges(opts);
+
+    publish.pubImpl(isTest, opts);
+};
+
+module.exports = publish;
+
+if (require.main === module) {
+
+    var opts = publish.getCLIOpts();
+    var isTest = opts["--test"];
+    var options = JSON.parse(opts["--options"] || "{}");
+
+    if (opts["--dev"]) {
```

Comment:
Perhaps it makes more sense to make the "dev" publication the default, rather than release? The primary value to the tool seems to be all of the workflow for making dev releases, since for real releases the user may as well just use "npm publish". The default operation of issuing a dev release also seems safer than producing a full one.